execution/src/vm.rs

```diff
@@ -0,0 +1,298 @@
+fn call_address(shared_env: &SharedExecutionContext, addr: Address) {
+    //if let Some(ledger_entry) = Arc::clone(&shared_env.0).lock().active_ledger.get(&addr) {}
+    //TODO
+}
+
+#[derive(Clone)]
+pub struct ExecutionStep {
+    pub slot: Slot,
+    // TODO add pos_seed for RNG seeding
+    // TODO add pos_draws to list the draws (block and endorsement creators) for that slot
+    pub block: Option<(BlockId, Block)>, // None if miss
+}
+
+#[derive(Clone)]
+/// Stateful context, providing an execution context to host functions("syscalls").
+pub struct ExecutionContext {
+    pub ledger_step: SCELedgerStep,
+    pub max_gas: u64,
+    pub coins: Amount,
+    pub gas_price: Amount,
+    pub slot: Slot,
+    pub opt_block_id: Option<BlockId>,
+    pub opt_block_creator_addr: Option<Address>,
+    pub call_stack: Vec<Address>,
+}
+
+#[derive(WasmerEnv, Clone)]
+pub struct SharedExecutionContext(pub Arc<Mutex<ExecutionContext>>);
+
+pub struct VM {
+    cfg: ExecutionConfig,
+    imports: ImportObject,
+    store: Store,
+    final_ledger: Arc<Mutex<SCELedger>>,
+    step_history: VecDeque<(Slot, Option<BlockId>, SCELedgerChanges)>,
+    current_execution_context: SharedExecutionContext,
+}
+
+impl VM {
+    pub fn new(cfg: ExecutionConfig) -> VM {
+        let store = Store::default();
+        let final_ledger = Arc::new(Mutex::new(
+            SCELedger::default(), // TODO bootstrap
+        ));
+        let current_execution_context =
+            SharedExecutionContext(Arc::new(Mutex::new(ExecutionContext {
+                ledger_step: SCELedgerStep {
+                    final_ledger: final_ledger.clone(),
+                    cumulative_history_changes: Default::default(),
+                    caused_changes: Default::default(),
+                },
+                max_gas: Default::default(),
+                coins: Default::default(),
+                gas_price: Default::default(),
+                slot: Slot::new(0, 0),
+                opt_block_id: Default::default(),
+                opt_block_creator_addr: Default::default(),
+                call_stack: Default::default(),
+            })));
+        let imports = imports! {
+            "env" => {
+                "foo" => Function::new_native_with_env(&store, current_execution_context.clone(), foo),
+            },
+        };
+        VM {
+            cfg,
+            imports,
+            store,
+            final_ledger,
+            step_history: Default::default(),
+            current_execution_context,
+        }
+    }
+
+    /// runs an SCE-final execution step
+    /// # Parameters
+    ///   * step: execution step to run
+    pub fn run_final_step(&mut self, step: ExecutionStep) {
+        // check if step already in history front
+        let opt_cached = {
+            if let Some((slot, opt_block, ledger_changes)) = self.step_history.pop_front() {
+                if slot != step.slot {
+                    // slot mismatch
```

Comment:
this one should not happen, but just in case there is a desync, better be robust. 

However, the failure cases listed just after can happen if there is a massive blockclique + finality change at a given time